### ISSUE TITLE
PulseAudio hangup on Raspberry Pi 3B

### ISSUE THREAD

**Issue by gavv** [roc-project/roc#135]
roc-recv sometimes hangs up on PulseAudio write indefinitely.

```
$ ./roc-recv -vv -s :10001 -r :10002 --resampler-frame=128 --resampler-window=32

(gdb) bt
#0  0x76f07a70 in __pthread_cond_wait (cond=0x2893c8, mutex=0x2893a8) at pthread_cond_wait.c:177
#1  0x76ec53a4 in pa_threaded_mainloop_wait () from /usr/lib/arm-linux-gnueabihf/libpulse.so.0
#2  0x76ee8908 in pa_simple_write () from /usr/lib/arm-linux-gnueabihf/libpulse-simple.so.0
#3  0x0008534c in write_samples ()
#4  0x0005ca90 in sox_write ()
#5  0x00011e78 in roc::sndio::Player::write_(int const*, unsigned int) ()
#6  0x00012038 in roc::sndio::Player::loop_() ()
#7  0x7e942b40 in ?? ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

```

**Comment by gavv** [comment]
```
I: [pulseaudio] protocol-native.c: Requested tlength=250.00 ms, minreq=20.00 ms
D: [pulseaudio] protocol-native.c: Adjust latency mode enabled, configuring sink latency to half of overall latency.
D: [pulseaudio] protocol-native.c: Requested latency=105.00 ms, Received latency=105.00 ms
D: [pulseaudio] memblockq.c: memblockq requested: maxlength=4194304, tlength=55680, base=8, prebuf=48008, minreq=7680 maxrewind=0
D: [pulseaudio] memblockq.c: memblockq sanitized: maxlength=4194304, tlength=55680, base=8, prebuf=48008, minreq=7680 maxrewind=0
I: [pulseaudio] protocol-native.c: Final latency 250.00 ms = 105.00 ms + 2*20.00 ms + 105.00 ms
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Latency set to 105.00ms
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: hwbuf_unused=110912
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: setting avail_min=31887
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Requesting rewind due to latency change.
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Requested volume: front-left: 59065 /  90% / -2.71 dB,   front-right: 59065 /  90% / -2.71 dB
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Got hardware volume: front-left: 59063 /  90% / -2.71 dB,   front-right: 59063 /  90% / -2.71 dB
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Calculated software volume: front-left: 65538 / 100% / 0.00 dB,   front-right: 65538 / 100% / 0.00 dB (accurate-enough=yes)
D: [alsa-sink-bcm2835 ALSA] sink.c: Volume not changing
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Requested to rewind 131072 bytes.
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Limited to 126452 bytes.
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: before: 31613
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: after: 31613
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Rewound 126452 bytes.
D: [alsa-sink-bcm2835 ALSA] sink.c: Processing rewind...
D: [alsa-sink-bcm2835 ALSA] sink.c: latency = 0
D: [alsa-sink-bcm2835 ALSA] sink-input.c: Have to rewind 126452 bytes on render memblockq.
D: [alsa-sink-bcm2835 ALSA] source.c: Processing rewind...
D: [alsa-sink-bcm2835 ALSA] protocol-native.c: Requesting rewind due to end of underrun.
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Requested to rewind 131072 bytes.
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Limited to 19904 bytes.
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: before: 4976
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: after: 4976
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Rewound 19904 bytes.
D: [alsa-sink-bcm2835 ALSA] sink.c: Processing rewind...
D: [alsa-sink-bcm2835 ALSA] sink.c: latency = 0
D: [alsa-sink-bcm2835 ALSA] sink-input.c: Have to rewind 19904 bytes on render memblockq.
D: [alsa-sink-bcm2835 ALSA] source.c: Processing rewind...
```

**Comment by gavv** [comment]
```
I: [pulseaudio] protocol-native.c: Requested tlength=250.00 ms, minreq=20.00 ms
D: [pulseaudio] protocol-native.c: Adjust latency mode enabled, configuring sink latency to half of overall latency.
D: [pulseaudio] protocol-native.c: Requested latency=105.00 ms, Received latency=105.00 ms
D: [pulseaudio] memblockq.c: memblockq requested: maxlength=4194304, tlength=55680, base=8, prebuf=48008, minreq=7680 maxrewind=0
D: [pulseaudio] memblockq.c: memblockq sanitized: maxlength=4194304, tlength=55680, base=8, prebuf=48008, minreq=7680 maxrewind=0
I: [pulseaudio] protocol-native.c: Final latency 250.00 ms = 105.00 ms + 2*20.00 ms + 105.00 ms
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Latency set to 105.00ms
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: hwbuf_unused=110912
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: setting avail_min=32568
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Requesting rewind due to latency change.
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Requested volume: front-left: 59065 /  90% / -2.71 dB,   front-right: 59065 /  90% / -2.71 dB
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Got hardware volume: front-left: 59063 /  90% / -2.71 dB,   front-right: 59063 /  90% / -2.71 dB
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Calculated software volume: front-left: 65538 / 100% / 0.00 dB,   front-right: 65538 / 100% / 0.00 dB (accurate-enough=yes)
D: [alsa-sink-bcm2835 ALSA] sink.c: Volume not changing
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Requested to rewind 131072 bytes.
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Limited to 130816 bytes.
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: before: 32704
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: after: 32704
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Rewound 130816 bytes.
D: [alsa-sink-bcm2835 ALSA] sink.c: Processing rewind...
D: [alsa-sink-bcm2835 ALSA] sink.c: latency = 0
D: [alsa-sink-bcm2835 ALSA] sink-input.c: Have to rewind 130816 bytes on render memblockq.
D: [alsa-sink-bcm2835 ALSA] source.c: Processing rewind...
D: [alsa-sink-bcm2835 ALSA] protocol-native.c: Requesting rewind due to end of underrun.
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Requested to rewind 131072 bytes.
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Limited to 19904 bytes.
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: before: 4976
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: after: 4976
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Rewound 19904 bytes.
D: [alsa-sink-bcm2835 ALSA] sink.c: Processing rewind...
D: [alsa-sink-bcm2835 ALSA] sink.c: latency = 0
D: [alsa-sink-bcm2835 ALSA] sink-input.c: Have to rewind 19904 bytes on render memblockq.
D: [alsa-sink-bcm2835 ALSA] source.c: Processing rewind...
I: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Underrun!
I: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Increasing wakeup watermark to 8.38 ms
D: [alsa-sink-bcm2835 ALSA] protocol-native.c: Implicit underrun of 'playback'
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 5312 bytes ago (14848 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 5312 bytes ago (14848 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 5312 bytes ago (14848 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 5312 bytes ago (12932 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 5312 bytes ago (12932 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 5312 bytes ago (9092 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 5312 bytes ago (9092 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 14920 bytes ago (5240 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] protocol-native.c: Requesting rewind due to end of underrun.
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Requested to rewind 14920 bytes.
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Limited to 14920 bytes.
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: before: 3730
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: after: 3730
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Rewound 14920 bytes.
D: [alsa-sink-bcm2835 ALSA] sink.c: Processing rewind...
D: [alsa-sink-bcm2835 ALSA] sink.c: latency = 0
D: [alsa-sink-bcm2835 ALSA] sink-input.c: Have to rewind 14920 bytes on render memblockq.
D: [alsa-sink-bcm2835 ALSA] source.c: Processing rewind...
D: [alsa-sink-bcm2835 ALSA] protocol-native.c: Implicit underrun of 'playback'
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 192 bytes ago (19968 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 192 bytes ago (19968 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 192 bytes ago (19968 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 192 bytes ago (18052 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 192 bytes ago (18052 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 192 bytes ago (14212 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 192 bytes ago (14212 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 9852 bytes ago (10308 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] protocol-native.c: Requesting rewind due to end of underrun.
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Requested to rewind 9852 bytes.
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Limited to 9852 bytes.
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: before: 2463
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: after: 2463
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Rewound 9852 bytes.
D: [alsa-sink-bcm2835 ALSA] sink.c: Processing rewind...
D: [alsa-sink-bcm2835 ALSA] sink.c: latency = 0
D: [alsa-sink-bcm2835 ALSA] sink-input.c: Have to rewind 9852 bytes on render memblockq.
D: [alsa-sink-bcm2835 ALSA] source.c: Processing rewind...
D: [alsa-sink-bcm2835 ALSA] protocol-native.c: Implicit underrun of 'playback'
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 88 bytes ago (20072 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 88 bytes ago (20072 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 88 bytes ago (20072 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 88 bytes ago (20072 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 88 bytes ago (18308 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 88 bytes ago (18308 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 88 bytes ago (14456 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 88 bytes ago (14456 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 9540 bytes ago (10620 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] protocol-native.c: Requesting rewind due to end of underrun.
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Requested to rewind 9540 bytes.
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Limited to 9540 bytes.
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: before: 2385
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: after: 2385
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Rewound 9540 bytes.
D: [alsa-sink-bcm2835 ALSA] sink.c: Processing rewind...
D: [alsa-sink-bcm2835 ALSA] sink.c: latency = 0
D: [alsa-sink-bcm2835 ALSA] sink-input.c: Have to rewind 9540 bytes on render memblockq.
D: [alsa-sink-bcm2835 ALSA] source.c: Processing rewind...
D: [alsa-sink-bcm2835 ALSA] protocol-native.c: Implicit underrun of 'playback'
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 1020 bytes ago (19140 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 1020 bytes ago (19140 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 1020 bytes ago (19140 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 1020 bytes ago (17408 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 1020 bytes ago (17408 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 1020 bytes ago (13572 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 1020 bytes ago (13572 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] sink.c: alsa_output.platform-soc_audio.analog-stereo: Found underrun 10820 bytes ago (9340 bytes ahead in playback buffer)
D: [alsa-sink-bcm2835 ALSA] protocol-native.c: Requesting rewind due to end of underrun.
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Requested to rewind 10820 bytes.
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Limited to 10820 bytes.
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: before: 2705
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: after: 2705
D: [alsa-sink-bcm2835 ALSA] alsa-sink.c: Rewound 10820 bytes.
D: [alsa-sink-bcm2835 ALSA] sink.c: Processing rewind...
D: [alsa-sink-bcm2835 ALSA] sink.c: latency = 0
D: [alsa-sink-bcm2835 ALSA] sink-input.c: Have to rewind 10820 bytes on render memblockq.
D: [alsa-sink-bcm2835 ALSA] source.c: Processing rewind...
```

**Comment by gavv** [comment]
Likely a bug in pulseaudio. It requires a combination of an underrun and high CPU usage. I can reproduces this without Roc and SoX: https://gist.github.com/gavv/9175515a9c3fb2539713212417c66a63

This is reproduced on pulseaudio 10.0, but seems to be fixed in pulseaudio 11.1.

**Comment by gavv** [comment]
Reliably reproduced on pulseaudio 10.0 and pulseaudio 11.1, e.g.:

```
./a.out 10000 5
```

or:

```
./a.out 100000 5
```

**Comment by gavv** [comment]
https://lists.freedesktop.org/archives/pulseaudio-discuss/2018-January/029253.html

**Comment by gavv** [comment]
Will be fixed by #149.

**Comment by gavv** [comment]
Fixed. We now use custom PulseAudio sink that detects hangups and reconnects to PulseAudio. BTW, these hangups were reproduced on PC as well, on a virtualbox running Linux.
